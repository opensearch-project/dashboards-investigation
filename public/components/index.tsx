/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import ReactDOM from 'react-dom';
import { AppMountParameters } from '../../../../src/core/public';
import { NoteBookServices } from '../types';
import { App } from './app';
import { OpenSearchDashboardsContextProvider } from '../../../../src/plugins/opensearch_dashboards_react/public';

export const Observability = (
  services: NoteBookServices,
  AppMountParametersProp: AppMountParameters
) => {
  ReactDOM.render(
    <OpenSearchDashboardsContextProvider services={services}>
      <App />
    </OpenSearchDashboardsContextProvider>,
    AppMountParametersProp.element
  );

  return () => ReactDOM.unmountComponentAtNode(AppMountParametersProp.element);
};
