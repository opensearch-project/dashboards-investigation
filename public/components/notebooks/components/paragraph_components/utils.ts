/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { NotebookType } from '../../../../../common/types/notebooks';

/**
 * Checks if this paragraph is in an agentic notebook and has been run before
 * @param notebookType - The type of notebook
 * @param paragraphIndex - The index of the current paragraph
 * @param totalParagraphs - The total number of paragraphs in the notebook
 * @returns true if the paragraph is agentic and has been run before (not the last paragraph)
 */
export const isAgenticRunBefore = (
  notebookType: NotebookType | undefined,
  paragraphIndex: number,
  totalParagraphs: number
): boolean => {
  return notebookType === NotebookType.AGENTIC && paragraphIndex < totalParagraphs - 1;
};
