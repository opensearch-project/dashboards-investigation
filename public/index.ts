/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import './components/notebooks/index.scss';
import { InvestigationPlugin } from './plugin';
import './variables.scss';

export { InvestigationPlugin as Plugin };

export const plugin = () => new InvestigationPlugin();

export { InvestigationStart } from './types';
