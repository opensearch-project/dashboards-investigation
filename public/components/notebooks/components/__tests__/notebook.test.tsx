/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import '@testing-library/jest-dom';
import { act, fireEvent, render, waitFor } from '@testing-library/react';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import { HttpResponse } from '../../../../../../../src/core/public';
import { getOSDHttp } from '../../../../../common/utils';
import {
  addQueryResponse,
  clearOutputNotebook,
  codeBlockNotebook,
  codePlaceholderText,
  emptyNotebook,
  migrateBlockNotebook,
  notebookPutResponse,
  runCodeBlockResponse,
} from '../../../../../test/notebooks_constants';
import { Notebook, NotebookProps } from '../notebook';
import {
  chromeServiceMock,
  notificationServiceMock,
  savedObjectsServiceMock,
} from '../../../../../../../src/core/public/mocks';
import { NOTEBOOKS_API_PREFIX } from '../../../../../common/constants/notebooks';
import { OpenSearchDashboardsContextProvider } from '../../../../../../../src/plugins/opensearch_dashboards_react/public';

jest.mock('../../../../../../../src/plugins/embeddable/public', () => ({
  ViewMode: {
    EDIT: 'edit',
    VIEW: 'view',
  },
}));

jest.mock('react-router-dom', () => ({
  useLocation: jest.fn().mockReturnValue({
    pathname: '/notebooks',
    search: '',
    hash: '',
    state: null,
    key: '',
  }),
  useHistory: jest.fn().mockReturnValue({
    replace: jest.fn(),
    push: jest.fn(),
  }),
}));

jest.mock('../data_distribution/data_distribution_container', () => ({
  DataDistributionContainer: () => <div />,
}));

jest.mock('../alert_panel', () => ({
  AlertPanel: () => <div />,
}));

jest.mock('../input_panel.tsx', () => ({
  InputPanel: () => <div />,
}));

// @ts-ignore
global.fetch = jest.fn(() =>
  Promise.resolve({
    json: () =>
      Promise.resolve({
        status: {
          statuses: [{ id: 'plugin:reportsDashboards' }],
        },
      }),
  })
);

const mockBehaviorSubject = jest.fn().mockImplementation(() => ({
  getValue: jest.fn(),
  subscribe: jest.fn(),
  next: jest.fn(),
}));

jest.mock('../../../../../public/services', () => ({
  getDataSourceManagementSetup: jest.fn(() => ({
    dataSourceManagement: {
      ui: {
        DataSourceSelector: () => <div />,
        getDataSourceMenu: jest.fn(),
      },
      registerAuthenticationMethod: jest.fn(),
      dataSourceSelection: {
        selectedDataSource$: new mockBehaviorSubject(),
        removedComponentIds: [],
        selectDataSource: jest.fn(),
        remove: jest.fn(),
        getSelectionValue: jest.fn(),
        getSelection$: () => new mockBehaviorSubject(),
      },
      getDefaultDataSourceId: jest.fn(),
      getDefaultDataSourceId$: jest.fn(() => new mockBehaviorSubject()),
    },
  })),
}));

const ContextAwareNotebook = (props: NotebookProps & { dataSourceEnabled?: boolean }) => {
  return (
    <OpenSearchDashboardsContextProvider
      services={{
        http: getOSDHttp(),
        dashboard: {
          DashboardContainerByValueRenderer: jest.fn(),
        },
        dataSource: props.dataSourceEnabled ? {} : undefined,
        chrome: chromeServiceMock.createStartContract(),
        savedObjects: savedObjectsServiceMock.createStartContract(),
        notifications: notificationServiceMock.createStartContract(),
      }}
    >
      <Notebook {...props} />
    </OpenSearchDashboardsContextProvider>
  );
};

describe('<Notebook /> spec', () => {
  configure({ adapter: new Adapter() });
  const httpClient = getOSDHttp();
  const location = jest.fn() as any;
  location.search = '';
  const history = jest.fn() as any;
  history.replace = jest.fn();
  history.push = jest.fn();
  const defaultProps: NotebookProps = {
    openedNoteId: '458e1320-3f05-11ef-bd29-e58626f102c0',
  };

  it('Renders the empty component', async () => {
    httpClient.get = jest.fn(() => Promise.resolve((emptyNotebook as unknown) as HttpResponse));
    const utils = render(<ContextAwareNotebook {...defaultProps} />);
    await waitFor(() => {
      expect(utils.getByText('sample-notebook-1')).toBeInTheDocument();
    });
    expect(utils.container.firstChild).toMatchSnapshot();
  });

  it('Adds a query', async () => {
    httpClient.get = jest.fn(() => Promise.resolve((emptyNotebook as unknown) as HttpResponse));
    let postFlag = 1;
    httpClient.post = jest.fn(() => {
      if (postFlag === 1) {
        postFlag += 1;
        return Promise.resolve((addQueryResponse as unknown) as HttpResponse);
      } else return Promise.resolve((runCodeBlockResponse as unknown) as HttpResponse);
    });
    const utils = render(<ContextAwareNotebook {...defaultProps} />);
    await waitFor(() => {
      expect(utils.getByText('sample-notebook-1')).toBeInTheDocument();
    });

    act(() => {
      utils.getByText('Add query').click();
    });

    await waitFor(() => {
      expect(utils.getByPlaceholderText(codePlaceholderText)).toBeInTheDocument();
    });
  });

  it('runs a query and checks the output', async () => {
    httpClient.get = jest.fn(() => Promise.resolve((emptyNotebook as unknown) as HttpResponse));
    let postFlag = 1;
    httpClient.post = jest.fn(() => {
      if (postFlag === 1) {
        postFlag += 1;
        return Promise.resolve((addQueryResponse as unknown) as HttpResponse);
      } else return Promise.resolve((runCodeBlockResponse as unknown) as HttpResponse);
    });
    const utils = render(<ContextAwareNotebook {...defaultProps} />);
    await waitFor(() => {
      expect(utils.getByText('sample-notebook-1')).toBeInTheDocument();
    });

    act(() => {
      utils.getByText('Add query').click();
    });

    await waitFor(() => {
      expect(utils.getByPlaceholderText(codePlaceholderText)).toBeInTheDocument();
    });

    act(() => {
      fireEvent.input(utils.getByPlaceholderText(codePlaceholderText), {
        target: { value: '%md \\n hello' },
      });
    });

    act(() => {
      // wait sometime to rerender
    });

    act(() => {
      fireEvent.click(utils.getByText('Run'));
    });

    await waitFor(() => {
      expect(utils.getByText('hello')).toBeInTheDocument();
    });
  });

  it('Checks notebook rename action', async () => {
    httpClient.get = jest.fn(() => Promise.resolve((codeBlockNotebook as unknown) as HttpResponse));

    httpClient.put = jest.fn(() => {
      return Promise.resolve((notebookPutResponse as unknown) as HttpResponse);
    });

    httpClient.post = jest.fn(() => {
      return Promise.resolve((addQueryResponse as unknown) as HttpResponse);
    });

    const utils = render(<ContextAwareNotebook {...defaultProps} />);
    await waitFor(() => {
      expect(utils.getByText('sample-notebook-1')).toBeInTheDocument();
    });

    act(() => {
      fireEvent.click(utils.getByTestId('notebook-edit-icon'));
    });

    await waitFor(() => {
      expect(utils.queryByTestId('custom-input-modal-input')).toBeInTheDocument();
    });

    act(() => {
      fireEvent.input(utils.getByTestId('custom-input-modal-input'), {
        target: { value: 'test-notebook-newname' },
      });
      fireEvent.click(utils.getByTestId('custom-input-modal-confirm-button'));
    });

    await waitFor(() => {
      expect(httpClient.put).toHaveBeenCalledWith(
        `${NOTEBOOKS_API_PREFIX}/note/savedNotebook/rename`,
        {
          body: JSON.stringify({
            name: 'sample-notebook-1',
            noteId: '458e1320-3f05-11ef-bd29-e58626f102c0',
          }),
        }
      );
    });
  });

  it('Checks notebook clone action', async () => {
    httpClient.get = jest.fn(() => Promise.resolve((codeBlockNotebook as unknown) as HttpResponse));

    httpClient.put = jest.fn(() => {
      return Promise.resolve((notebookPutResponse as unknown) as HttpResponse);
    });

    httpClient.post = jest.fn(() => {
      return Promise.resolve((addQueryResponse as unknown) as HttpResponse);
    });

    const utils = render(<ContextAwareNotebook {...defaultProps} />);
    await waitFor(() => {
      expect(utils.getByText('sample-notebook-1')).toBeInTheDocument();
    });

    act(() => {
      fireEvent.click(utils.getByTestId('notebook-duplicate-icon'));
    });

    await waitFor(() => {
      expect(utils.queryByTestId('custom-input-modal-input')).toBeInTheDocument();
    });

    act(() => {
      fireEvent.click(utils.getByTestId('custom-input-modal-confirm-button'));
    });

    expect(httpClient.post).toHaveBeenCalledWith(
      `${NOTEBOOKS_API_PREFIX}/note/savedNotebook/clone`,
      {
        body: JSON.stringify({
          name: 'sample-notebook-1 (copy)',
          noteId: '458e1320-3f05-11ef-bd29-e58626f102c0',
        }),
      }
    );
  });

  it('Checks notebook delete action', async () => {
    httpClient.get = jest.fn(() => Promise.resolve((codeBlockNotebook as unknown) as HttpResponse));

    httpClient.put = jest.fn(() => {
      return Promise.resolve((notebookPutResponse as unknown) as HttpResponse);
    });

    httpClient.post = jest.fn(() => {
      return Promise.resolve((addQueryResponse as unknown) as HttpResponse);
    });

    const utils = render(<ContextAwareNotebook {...defaultProps} />);
    await waitFor(() => {
      expect(utils.getByText('sample-notebook-1')).toBeInTheDocument();
    });

    act(() => {
      fireEvent.click(utils.getByTestId('notebook-delete-icon'));
    });

    await waitFor(() => {
      expect(utils.queryByTestId('delete-notebook-modal-input')).toBeInTheDocument();
    });

    act(() => {
      fireEvent.input(utils.getByTestId('delete-notebook-modal-input'), {
        target: { value: 'delete' },
      });
    });

    act(() => {
      fireEvent.click(utils.getByTestId('delete-notebook-modal-delete-button'));
    });

    expect(httpClient.delete).toHaveBeenCalledWith(
      `${NOTEBOOKS_API_PREFIX}/note/savedNotebook/458e1320-3f05-11ef-bd29-e58626f102c0`
    );
  });

  it('Checks notebook reporting action presence', async () => {
    httpClient.get = jest.fn(() => Promise.resolve((emptyNotebook as unknown) as HttpResponse));

    const utils = render(<ContextAwareNotebook {...defaultProps} />);
    await waitFor(() => {
      expect(utils.getByText('sample-notebook-1')).toBeInTheDocument();
    });

    const button = utils.queryByTestId('reporting-actions-button');
    expect(button).toBeInTheDocument();
  });

  it('Checks notebook reporting action absence', async () => {
    httpClient.get = jest.fn(() => Promise.resolve((emptyNotebook as unknown) as HttpResponse));

    const utils = render(<ContextAwareNotebook {...defaultProps} dataSourceEnabled />);
    await waitFor(() => {
      expect(utils.getByText('sample-notebook-1')).toBeInTheDocument();
    });
    const button = utils.queryByTestId('reporting-actions-button');
    expect(button).not.toBeInTheDocument();
  });

  it('Renders a old notebook and migrates it', async () => {
    httpClient.get = jest.fn(() => Promise.resolve((codeBlockNotebook as unknown) as HttpResponse));
    httpClient.put = jest.fn(() =>
      Promise.resolve((clearOutputNotebook as unknown) as HttpResponse)
    );
    httpClient.delete = jest.fn(() =>
      Promise.resolve(({ paragraphs: [] } as unknown) as HttpResponse)
    );
    httpClient.get = jest.fn(() =>
      Promise.resolve((migrateBlockNotebook as unknown) as HttpResponse)
    );
    const utils = render(<ContextAwareNotebook {...defaultProps} openedNoteId="mock-id" />);
    await waitFor(() => {
      expect(
        utils.getByText('Upgrade this notebook to take full advantage of the latest features')
      ).toBeInTheDocument();
    });

    act(() => {
      fireEvent.click(utils.getByTestId('upgrade-notebook'));
    });

    act(() => {
      fireEvent.click(utils.getByTestId('custom-input-modal-confirm-button'));
    });

    expect(httpClient.post).toHaveBeenCalledWith(`${NOTEBOOKS_API_PREFIX}/note/migrate`, {
      body: JSON.stringify({
        name: 'sample-notebook-1 (upgraded)',
        noteId: 'mock-id',
      }),
    });
  });

  it('Checks old notebook delete action', async () => {
    httpClient.get = jest.fn(() => Promise.resolve((codeBlockNotebook as unknown) as HttpResponse));

    httpClient.put = jest.fn(() => {
      return Promise.resolve((notebookPutResponse as unknown) as HttpResponse);
    });

    httpClient.post = jest.fn(() => {
      return Promise.resolve((addQueryResponse as unknown) as HttpResponse);
    });

    const utils = render(<ContextAwareNotebook {...defaultProps} openedNoteId="mock-id" />);
    await waitFor(() => {
      expect(utils.getByText('sample-notebook-1')).toBeInTheDocument();
    });

    act(() => {
      fireEvent.click(utils.getByTestId('notebook-delete-icon'));
    });

    await waitFor(() => {
      expect(utils.queryByTestId('delete-notebook-modal-input')).toBeInTheDocument();
    });

    act(() => {
      fireEvent.input(utils.getByTestId('delete-notebook-modal-input'), {
        target: { value: 'delete' },
      });
    });

    act(() => {
      fireEvent.click(utils.getByTestId('delete-notebook-modal-delete-button'));
    });

    expect(httpClient.delete).toHaveBeenCalledWith(`${NOTEBOOKS_API_PREFIX}/note/mock-id`);
  });
});
