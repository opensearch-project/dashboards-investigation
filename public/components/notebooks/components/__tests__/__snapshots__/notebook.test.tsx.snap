// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Notebook /> spec Renders the empty component 1`] = `
<div
  class="euiPage euiPage--paddingMedium euiPage--column euiPage--grow"
>
  <div
    class="euiPageBody euiPageBody--borderRadiusNone"
  >
    <div>
      <div
        class="euiFlexGroup euiFlexGroup--gutterSmall euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
      >
        <h3
          class="euiTitle euiTitle--large"
          data-test-subj="notebookTitle"
        >
          sample-notebook-1
        </h3>
        <div
          class="euiFlexItem euiFlexItem--flexGrowZero"
        >
          <div
            class="euiFlexGroup euiFlexGroup--gutterSmall euiFlexGroup--alignItemsCenter euiFlexGroup--directionRow euiFlexGroup--responsive"
          >
            <div
              class="euiFlexGroup euiFlexGroup--gutterSmall euiFlexGroup--directionRow euiFlexGroup--responsive"
            >
              <div
                class="euiFlexItem euiFlexItem--flexGrowZero"
              >
                <span
                  class="euiToolTipAnchor"
                >
                  <button
                    aria-label="Edit system prompt"
                    class="euiButtonIcon euiButtonIcon--primary euiButtonIcon--small"
                    data-test-subj="notebook-system-prompt-icon"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="euiIcon euiIcon--medium euiIcon--inherit euiIcon-isLoading euiButtonIcon__icon"
                      focusable="false"
                      height="16"
                      role="img"
                      viewBox="0 0 16 16"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
                      />
                    </svg>
                  </button>
                </span>
              </div>
              <div
                class="euiFlexItem euiFlexItem--flexGrowZero"
              >
                <span
                  class="euiToolTipAnchor"
                >
                  <button
                    aria-label="Delete"
                    class="euiButtonIcon euiButtonIcon--danger euiButtonIcon--small"
                    data-test-subj="notebook-delete-icon"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="euiIcon euiIcon--medium euiIcon--inherit euiIcon-isLoading euiButtonIcon__icon"
                      focusable="false"
                      height="16"
                      role="img"
                      viewBox="0 0 16 16"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
                      />
                    </svg>
                  </button>
                </span>
              </div>
              <div
                class="euiFlexItem euiFlexItem--flexGrowZero"
              >
                <span
                  class="euiToolTipAnchor"
                >
                  <button
                    aria-label="Edit name"
                    class="euiButtonIcon euiButtonIcon--primary euiButtonIcon--small"
                    data-test-subj="notebook-edit-icon"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="euiIcon euiIcon--medium euiIcon--inherit euiIcon-isLoading euiButtonIcon__icon"
                      focusable="false"
                      height="16"
                      role="img"
                      viewBox="0 0 16 16"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
                      />
                    </svg>
                  </button>
                </span>
              </div>
              <div
                class="euiFlexItem euiFlexItem--flexGrowZero"
              >
                <span
                  class="euiToolTipAnchor"
                >
                  <button
                    aria-label="Duplicate"
                    class="euiButtonIcon euiButtonIcon--primary euiButtonIcon--small"
                    data-test-subj="notebook-duplicate-icon"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="euiIcon euiIcon--medium euiIcon--inherit euiIcon-isLoading euiButtonIcon__icon"
                      focusable="false"
                      height="16"
                      role="img"
                      viewBox="0 0 16 16"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
                      />
                    </svg>
                  </button>
                </span>
              </div>
            </div>
            <div
              class="euiFlexItem euiFlexItem--flexGrowZero"
            >
              <div>
                <div
                  class="euiPopover euiPopover--anchorDownCenter"
                >
                  <div
                    class="euiPopover__anchor"
                  >
                    <button
                      class="euiButton euiButton--primary euiButton--small"
                      data-test-subj="reporting-actions-button"
                      id="reportingActionsButton"
                      type="button"
                    >
                      <span
                        class="euiButtonContent euiButtonContent--iconRight euiButton__content"
                      >
                        <svg
                          aria-hidden="true"
                          class="euiIcon euiIcon--medium euiIcon--inherit euiIcon-isLoading euiButtonContent__icon"
                          focusable="false"
                          height="16"
                          role="img"
                          viewBox="0 0 16 16"
                          width="16"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
                          />
                        </svg>
                        <span
                          class="euiButton__text"
                        >
                          Reporting
                        </span>
                      </span>
                    </button>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="euiFlexItem euiFlexItem--flexGrowZero"
            />
          </div>
        </div>
      </div>
      <div
        class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
      >
        <div
          class="euiFlexItem euiFlexItem--flexGrowZero"
        >
          <p>
            Created on 12/14/2023 06:49 PM
          </p>
        </div>
      </div>
      <div
        class="euiSpacer euiSpacer--s"
      />
    </div>
    <div
      class="euiSpacer euiSpacer--l"
    />
    <div
      class="euiPanel euiPanel--paddingLarge euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow euiPageContent euiPageContent--horizontalCenter"
      role="main"
      style="width: 900px;"
    >
      <div
        style="margin-top: 10px;"
      >
        <div
          class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
        >
          <div
            class="euiSpacer euiSpacer--xxl"
          />
          <div
            class="euiText euiText--medium"
          >
            <div
              class="euiTextAlign euiTextAlign--center"
            >
              <h2>
                No paragraphs
              </h2>
              <div
                class="euiText euiText--small"
              >
                Add a paragraph to compose your document or story. Notebooks now support two types of input:
              </div>
            </div>
          </div>
          <div
            class="euiSpacer euiSpacer--xl"
          />
          <div
            class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--justifyContentSpaceEvenly euiFlexGroup--directionRow euiFlexGroup--responsive"
          >
            <div
              class="euiFlexItem euiFlexItem--flexGrow2"
            />
            <div
              class="euiFlexItem euiFlexItem--flexGrow3"
            >
              <div
                class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow euiCard euiCard--centerAligned euiCard--hasIcon"
              >
                <div
                  class="euiCard__top"
                >
                  <svg
                    aria-hidden="true"
                    class="euiIcon euiIcon--xxLarge euiCard__icon"
                    focusable="false"
                    height="16"
                    role="img"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M2.414 8.036 4.89 10.51a.5.5 0 0 1-.707.708L1.354 8.389a.5.5 0 0 1 0-.707l2.828-2.828a.5.5 0 1 1 .707.707L2.414 8.036Zm8.768 2.474 2.475-2.474-2.475-2.475a.5.5 0 0 1 .707-.707l2.829 2.828a.5.5 0 0 1 0 .707l-2.829 2.829a.5.5 0 1 1-.707-.708ZM8.559 2.506a.5.5 0 0 1 .981.19L7.441 13.494a.5.5 0 0 1-.981-.19L8.559 2.506Z"
                    />
                  </svg>
                </div>
                <div
                  class="euiCard__content"
                >
                  <span
                    class="euiTitle euiTitle--small euiCard__title"
                    id="random_html_idTitle"
                  >
                    Query
                  </span>
                  <div
                    class="euiText euiText--small euiCard__description"
                    id="random_html_idDescription"
                  >
                    <p>
                      Write contents directly using markdown, SQL or PPL.
                    </p>
                  </div>
                </div>
                <div
                  class="euiCard__footer"
                >
                  <button
                    class="euiButton euiButton--primary euiButton--small"
                    data-test-subj="emptyNotebookAddCodeBlockBtn"
                    style="margin-bottom: 17px;"
                    type="button"
                  >
                    <span
                      class="euiButtonContent euiButton__content"
                    >
                      <span
                        class="euiButton__text"
                      >
                        Add query
                      </span>
                    </span>
                  </button>
                </div>
              </div>
            </div>
            <div
              class="euiFlexItem euiFlexItem--flexGrow3"
            >
              <div
                class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow euiCard euiCard--centerAligned euiCard--hasIcon"
              >
                <div
                  class="euiCard__top"
                >
                  <svg
                    aria-hidden="true"
                    class="euiIcon euiIcon--xxLarge euiCard__icon"
                    focusable="false"
                    height="16"
                    role="img"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M3 13h10V9.913l-2.571-2.826L8.56 8.753a.5.5 0 0 1-.728-.067L4.448 4.317 3 6.191V13Zm5.295-5.35 1.837-1.64a.5.5 0 0 1 .703.037l3.035 3.336a.5.5 0 0 1 .13.337v3.78a.5.5 0 0 1-.5.5h-11a.5.5 0 0 1-.5-.5V6.02a.5.5 0 0 1 .104-.305l1.947-2.52a.5.5 0 0 1 .791-.001L8.295 7.65ZM1 15h13.5a.5.5 0 1 1 0 1H.5a.5.5 0 0 1-.5-.5v-14a.5.5 0 0 1 1 0V15Z"
                    />
                  </svg>
                </div>
                <div
                  class="euiCard__content"
                >
                  <span
                    class="euiTitle euiTitle--small euiCard__title"
                    id="random_html_idTitle"
                  >
                    Visualization
                  </span>
                  <div
                    class="euiText euiText--small euiCard__description"
                    id="random_html_idDescription"
                  >
                    <p>
                      Import OpenSearch Dashboards or Observability visualizations to the notes.
                    </p>
                  </div>
                </div>
                <div
                  class="euiCard__footer"
                >
                  <button
                    class="euiButton euiButton--primary euiButton--small"
                    style="margin-bottom: 17px;"
                    type="button"
                  >
                    <span
                      class="euiButtonContent euiButton__content"
                    >
                      <span
                        class="euiButton__text"
                      >
                        Add visualization
                      </span>
                    </span>
                  </button>
                </div>
              </div>
            </div>
            <div
              class="euiFlexItem euiFlexItem--flexGrow3"
            >
              <div
                class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow euiCard euiCard--centerAligned euiCard--hasIcon"
              >
                <div
                  class="euiCard__top"
                >
                  <svg
                    aria-hidden="true"
                    class="euiIcon euiIcon--xxLarge euiCard__icon"
                    focusable="false"
                    height="16"
                    role="img"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M15.363 14.658a.5.5 0 1 1-.713.7l-2.97-3.023a.5.5 0 0 1 .001-.7A3.9 3.9 0 1 0 8.9 12.8a.5.5 0 1 1 0 .999 4.9 4.9 0 1 1 3.821-1.833l2.642 2.691ZM3.094 13a.5.5 0 1 1 0 1H2.5A2.5 2.5 0 0 1 0 11.5v-9A2.5 2.5 0 0 1 2.5 0h9A2.5 2.5 0 0 1 14 2.5v.599a.5.5 0 1 1-1 0V2.5A1.5 1.5 0 0 0 11.5 1h-9A1.5 1.5 0 0 0 1 2.5v9A1.5 1.5 0 0 0 2.5 13h.594ZM2.5 3a.5.5 0 1 1 0-1 .5.5 0 0 1 0 1Zm2 0a.5.5 0 1 1 0-1 .5.5 0 0 1 0 1Zm2 0a.5.5 0 1 1 0-1 .5.5 0 0 1 0 1Zm-4 2a.5.5 0 1 1 0-1 .5.5 0 0 1 0 1Zm2 0a.5.5 0 1 1 0-1 .5.5 0 0 1 0 1Zm-2 1a.5.5 0 1 1 0 1 .5.5 0 0 1 0-1Zm0 3a.5.5 0 1 1 0-1 .5.5 0 0 1 0 1Zm6-6a.5.5 0 1 1 0-1 .5.5 0 0 1 0 1Zm2 0a.5.5 0 1 1 0-1 .5.5 0 0 1 0 1Zm-8 8a.5.5 0 1 1 0-1 .5.5 0 0 1 0 1Z"
                    />
                  </svg>
                </div>
                <div
                  class="euiCard__content"
                >
                  <span
                    class="euiTitle euiTitle--small euiCard__title"
                    id="random_html_idTitle"
                  >
                    Deep Research
                  </span>
                  <div
                    class="euiText euiText--small euiCard__description"
                    id="random_html_idDescription"
                  >
                    <p>
                      Use deep research to analytics question.
                    </p>
                  </div>
                </div>
                <div
                  class="euiCard__footer"
                >
                  <button
                    class="euiButton euiButton--primary euiButton--small"
                    style="margin-bottom: 17px;"
                    type="button"
                  >
                    <span
                      class="euiButtonContent euiButton__content"
                    >
                      <span
                        class="euiButton__text"
                      >
                        Add deep research
                      </span>
                    </span>
                  </button>
                </div>
              </div>
            </div>
            <div
              class="euiFlexItem euiFlexItem--flexGrow2"
            />
          </div>
          <div
            class="euiSpacer euiSpacer--xxl"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    class="euiSpacer euiSpacer--l"
  />
  <div />
</div>
`;
